### ISSUE TITLE
Wasix stdin seems not to be working

### ISSUE THREAD

**Issue by Cactice** [wasmerio/wasmer#4146]
### Describe the bug

I'm porting mlir-translate to wasm and trying to run on wasmer_wasix.

When running on wasmtime_wasi, mlir-translate.wasm receives stdin and executes as expected.
However, when running equivalent code on wasmer_wasix, mlir-translate.wasm hangs.

mlir-translate typically hangs when stdin is not provided so I imagine stdin is not provided.

```sh
wasmer -vV; rustc -vV
wasmer 4.1.1 (0bd4b48 2023-08-03)
binary: wasmer-cli
commit-hash: 0bd4b48691924127200ea206f72b813e85d11520
commit-date: 2023-08-03
host: aarch64-apple-darwin
compiler: singlepass,cranelift
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: aarch64-apple-darwin
release: 1.71.1
LLVM version: 16.0.5
```

### Steps to reproduce

1. [Clone repo](https://github.com/Cactice/mlir-playground/tree/main)
1. cargo run --bin wasmer 
1. program hangs

### Expected behavior

Should log weird characters similar to wasmtime
```
➜  mlir-playground git:(main)  cargo run --bin wasmtime
called
read done
#2*maininout
combineaddFloatappendFloatvec3ToVec4*mainGG  ; ;        !        !!!+!�?+%@@)!()- ,-6
7
 7
P �

        �8677�==��867�>�867� Q"Q#Q$9&%P'"#$!�'86)*(�+;,.9       /
%%=0910>1�8% 
```

### Actual behavior

```
➜  mlir-playground git:(main)  cargo run --bin wasmer
Compiling module...
Done compile
Tokio ready
initialized
written
```
The program hangs here

**Comment by Michael-F-Bryan** [comment]
@Cactice you might need to drop the `stdin_sender` variable. This will close the STDIN stream so `mlir-translate` knows there is no more data to read.

**Comment by Cactice** [comment]
@Michael-F-Bryan 
Thank you so much for your quick response

The situation stdin seems fine but now reading stdout hangs...
I tried dropping wasi_env but it's not helping.
If you have an idea it'd be greatly helpful. Thanks!!

Changes: https://github.com/Cactice/mlir-playground/commit/44b066ee00463f49865b71aa6157927727e16eda

```
Compiling module...
Done compile
Tokio ready
initialized
written
called
reading...
```

**Comment by Michael-F-Bryan** [comment]
That makes sense because you are calling `_start()` directly, which won't do cleanup like closing handles and so on.

Instead of setting up imports, creating the `wasi_env`, and instantiating the module manually, I'd recommend using `WasiEnvBuilder::run()`.

```rust
let module = unsafe { Module::deserialize_from_file(&store, "./mlir-translate.wasm-module") }?;

let tokio_runtime = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
        .unwrap();
let _guard = runtime.enter();

let (mut stdin_sender, stdin_reader) = Pipe::channel();
let (stdout_sender, mut stdout_reader) = Pipe::channel();

println!("Starting");

WasiEnv::builder("mlir-translate")
        .stdin(Box::new(stdin_reader))
        .stdout(Box::new(stdout_sender))
        .arg("-serialize-spirv")
        .arg("-no-implicit-module")
        .arg("-mlir-disable-threading")
        .run(&module)?;

println!("mlir-translate completed successfully");

println!("Reading...");
let mut buf = Vec::new();
stdout_reader.read_to_end(&mut buf)?;

println!(
  "Read \"{}\" from the WASI stdout!",    
  String::from_utf8_lossy(&buf)
);
```

You may also need to read stdout in parallel. A `Pipe` will buffer data internally, but I don't remember how much data it'll buffer before writing to stdout inside the WASI instance blocks.

**Comment by Cactice** [comment]
@Michael-F-Bryan 
The issue with `run` is that it results in error like below 

```
Error while importing "env"."__syscall_faccessat": unknown import. Expected Function(FunctionType { params: [I32, I32, I32, I32], results: [I32] })
```

To avoid such error, I needed to import manually and couldn't use `run`.

Btw, I added  `wasi_env.cleanup(&mut store, None);` and now I can read stdout! (https://github.com/Cactice/mlir-playground/commit/6bc9cbb433e97894d2da86fee25db3e40fe91457)
Thank you so much, this is much faster than wasmtime, I appreciate it.
